### PR TITLE
Upload CI artifacts

### PR DESCRIPTION
**Pull request by Kha**
After @rwbarton mentioned that he had trouble building Lean 4 from source, it occurred to me that we could simply use Github Actions' support for publishing artifacts we're already building on CI. This revealed some issues we would need to address before an official release anyway.

The most frustrating issue is the glibc dependency on Linux: we can't link Lean statically anymore like we did with Lean 3 because our use of `dlopen`. But if we link it dynamically, it turns out that the executable does not run on any Linux machine with glibc < 2.29, which is still pretty new. The simplest solution I found is to use a separate build job for release that uses a Nix channel (19.03) with an older glibc (2.27), and sadly an older LLVM (7.0.1) as well. AFAICS, the resulting binary is actually compatible down to glibc 2.17, which should be plenty sufficient.

BTW, somewhat unexpectedly it turns out that you can only download artifacts while logged into Github, so this is definitely not an official release :sweat_smile:

### CODE REVIEW

**Comment by Kha** [comment]
I've checked that the "Linux release" binary works on Ubuntu 18.04 without Nix installed - on the other hand, I'm pretty sure that the macOS binary will not work as-is. An `install_name_tool -delete-rpath bin/lean` might fix that. @leodemoura Could you give this a try? The artifacts are available on the right hand side of the "Checks" tab.

**Comment by leodemoura** [comment]
@Kha I tried the macOS binary. It didn't work, there is a missing dylib (macOS shared library). Here is the error message:

```
Leos-MacBook-Pro:lean-4.0.0-darwin leonardodemoura$ bin/lean
dyld: Library not loaded: /nix/store/mazy2jylffgx0n90wfkjlppc3xmcb86b-libc++-9.0.1/lib/libc++.1.0.dylib
  Referenced from: /Users/leonardodemoura/code/lean-4.0.0-darwin/bin/lean
  Reason: image not found
Abort trap: 6
```

**Comment by Kha** [comment]
@leodemoura Could you do `install_name_tool -delete_rpath bin/lean` and try again?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -67,41 +71,58 @@ jobs:
...
+          make -j4
+      - name: Patch
+        run: patchelf --set-interpreter /lib64/ld-linux-x86-64.so.2 --remove-rpath build/stage0.5/bin/lean
+        if: matrix.name == 'Linux release'
+      - name: Patch
+        run: |
+          otool -l build/stage0.5/bin/lean
+          for lib in $(otool -l build/stage0.5/bin/lean | cut -d' ' -f1); do
+            install_name_tool -change "$lib" "/usr/lib/$(basename $lib)" build/stage0.5/bin/lean
+          done
```
Comment (Kha) [inline .github/workflows/ci.yml:95]: Huh, this might actually work. @leodemoura Could you try the current macOS artifact again?

**Pull request merged**